### PR TITLE
Apply tree shearing after matching table to tree

### PR DESCRIPTION
**Pull request by esayyari**
Addressing #191. Doing error checking using BP tree instead of TreeNode and doing tree shearing later in the code.

### CODE REVIEW

File: tests/python/test_core.py [hunk 1]
```diff
@@ -197,7 +209,19 @@ def test_to_dict_with_feature_metadata(self):
...
+        for node in tree_data:
+            for attr in tree_data[node]:
+                self.assertAlmostEqual(tree_data[node][attr], exp[node][attr])
+        # check that keys are identical otherwise
+        self.assertEqual(tree_data.keys(), exp.keys())
+
+        obs.pop('tree_data')
+        dict_a_with_fm.pop('tree_data')
         self.assertEqual(obs, dict_a_with_fm)
+        dict_a_with_fm['tree_data'] = exp
```
Comment (fedarko) [inline tests/python/test_core.py:224]: since this code is repeated three times within these tests, do you think it would be feasible to encapsulate most of this work to a separate utility function within this file? This would make writing additional core tests easier.

File: tests/python/test_tools.py [hunk 2]
```diff
@@ -80,7 +81,7 @@ def test_fill_missing_node_names(self):
             self.assertEqual(node.name, names[i])
 
     def test_match_inputs_nothing_dropped(self):
-        t = Tree.from_tree(self.tree)
+        t = from_skbio_treenode(self.tree)
```
Comment (fedarko) [inline tests/python/test_tools.py:84]: Rather than repeating this at the start of most of the tests here, it may be cleaner to define `self.bp_tree = from_skbio_treenode(self.tree)` in the `setUp()` function and then just use that for these functions.

(this is a minor issue, so feel free to ignore this.)

File: empress/core.py [hunk 3]
```diff
@@ -147,6 +138,15 @@ def _validate_and_match_data(self, ignore_missing_samples,
...
         )
+        # remove unobserved features from the phylogeny
+        if filter_unobserved_features_from_phylogeny:
+            self.tree = self.tree.shear(set(self.table.index))
+
+        # extract balance parenthesis
+        self._bp_tree = list(self.tree.B)
+
+        self.tree = Tree.from_tree(to_skbio_treenode(self.tree))
+        fill_missing_node_names(self.tree)
```
Comment (fedarko) [inline empress/core.py:149]: Thank you for addressing this!

If you don't mind, it would be nice to add a test case that verifies that the corner-case described in https://github.com/biocore/empress/pull/178#discussion_r441761697 has been addressed (so that this problem won't come up again). I think this would be doable by adding a test to the `test_core.py` file that uses a mismatched table and `filter_unobserved_features_from_phylogeny=True`, and verifies that a `tools.DataMatchingError` is raised (analogous to how the first few lines in [this unit test](https://github.com/biocore/empress/blob/148652b36345344eb8f7378265de434099eb77ec/tests/python/test_tools.py#L105-L113) work).

Comment (esayyari) [reply empress/core.py:149]: Needs to be addressed

File: tests/python/test_tools.py (hunk 2 above)
Comment (esayyari) [reply tests/python/test_tools.py]: Fixed

File: tests/python/test_core.py (hunk 1 above)
Comment (esayyari) [reply tests/python/test_core.py]: Created the method "assert_almost_equal_tree_data" to address this

File: tests/python/test_core.py [hunk 4]
```diff
@@ -181,7 +181,19 @@ def test_to_dict(self):
...
+        for node in tree_data:
+            for attr in tree_data[node]:
+                self.assertAlmostEqual(tree_data[node][attr], exp[node][attr])
+        # check that keys are identical otherwise
+        self.assertEqual(tree_data.keys(), exp.keys())
+
+        DICT_A.pop('tree_data')
+        obs.pop('tree_data')
         self.assertEqual(obs, DICT_A)
+        DICT_A['tree_data'] = exp
```
[the remark this answers is not in the document]
Comment (esayyari) [reply tests/python/test_core.py]: Fixed

File: empress/tree.py [hunk 5]
```diff
@@ -635,3 +635,23 @@ def update_unrooted_coords(self, s, x1, y1, a, da):
...
+def bp_tree_non_tips(bp_tree):
+    non_tips = []
+    for i in range(bp_tree.B.size):
+        if not isleaf(bp_tree, i):
+            non_tips.append(bp_tree.name(i))
+    return non_tips
+
+
+def isleaf(bp_tree, i):
+    return bp_tree.B[i] and (not bp_tree.B[i + 1])
```
[the remark this answers is not in the document]
Comment (esayyari) [reply empress/tree.py:697]: Added some documentation

File: empress/tools.py [hunk 6]
```diff
@@ -43,6 +45,32 @@ def fill_missing_node_names(tree):
...
+        can be None).
+
+    Returns
+    -------
+    skbio.TreeNode or empress.Tree
+        Tree with all nodes assigned a name.
+    """
+    current_unlabeled_node = 0
+    new_names = np.full(bp_tree.B.size, None, dtype=object)
+    for i in range(sum(bp_tree.B)):
```
[the remark this answers is not in the document]
Comment (esayyari) [reply empress/tools.py]: The function is removed.

File: empress/tools.py [hunk 7]
```diff
@@ -43,6 +45,32 @@ def fill_missing_node_names(tree):
...
+        if bp_tree.name(i) is None:
```
[the remark this answers is not in the document]
Comment (esayyari) [reply empress/tools.py]: The function is removed. Wasn't for this PR

File: empress/tools.py [hunk 8]
```diff
@@ -43,6 +45,32 @@ def fill_missing_node_names(tree):
             current_unlabeled_node += 1
 
 
+def fill_missing_node_names_from_bp_tree(bp_tree):
+    """ Names nodes in the tree without a name.
+
+     Parameters
+     ----------
+     tree : BP tree
```
[the remark this answers is not in the document]
Comment (esayyari) [reply empress/tools.py]: Fixed

File: empress/tools.py [hunk 9]
```diff
@@ -71,7 +99,7 @@ def match_inputs(
     Parameters
     ----------
 
-    tree: empress.tree.Tree
+    bp_tree: BP Tree
```
[the remark this answers is not in the document]
Comment (esayyari) [reply empress/tools.py]: Fixed

File: empress/tools.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fedarko) [reply empress/tools.py]: Got it, sorry for the confusion.

File: empress/tree.py [hunk 10]
```diff
@@ -635,3 +635,63 @@ def update_unrooted_coords(self, s, x1, y1, a, da):
...
+        Parameters
+       ----------
+       bp_tree : bp.BP
+           Input BP tree
+        i : int
+           The query node index
+       Returns
+       -------
+       int
+           1 if this is a leaf node, 0 otherwise
```
Comment (fedarko) [inline empress/tree.py:695]:
```suggestion
       bool
           True if this is a leaf node, False otherwise
```

(`and` in python returns a bool, not an integer)

File: empress/tree.py [hunk 11]
```diff
@@ -635,3 +635,63 @@ def update_unrooted_coords(self, s, x1, y1, a, da):
...
+    -------
+    tips : list of strings
+        list of tip names in the tree
+    """
+    tips = []
+    # Iterate through all open and closing parentheses and extract tip names
+    for i in range(bp_tree.B.size):
+        # Check if this is a leaf node and has a label
+        if isleaf(bp_tree, i) and (bp_tree.name(i) is not None):
+            tips.append(bp_tree.name(i))
```
Comment (fedarko) [inline empress/tree.py:657]:
```suggestion
        pos_name = bp_tree.name(i)
        # Check if this is a leaf node with a label
        if isleaf(bp_tree, i) and (pos_name is not None):
            tips.append(pos_name)
```

File: empress/tree.py [hunk 12]
```diff
@@ -635,3 +635,63 @@ def update_unrooted_coords(self, s, x1, y1, a, da):
...
+       non_tips : list of strings
+           list of internal node names in the tree
+    """
+    non_tips = []
+    for i in range(bp_tree.B.size):
+        # Check if this is an opening parenthesis, is not leaf, and
+        # has a node label
+        if bp_tree.B[i] and (not isleaf(bp_tree, i)) and \
+                (bp_tree.name(i) is not None):
+            non_tips.append(bp_tree.name(i))
```
Comment (fedarko) [inline empress/tree.py:679]:
```suggestion
        pos_name = bp_tree.name(i)
        # Check if this is an opening parenthesis, is not a leaf, and
        # has a node label
        if bp_tree.B[i] and not isleaf(bp_tree, i) and pos_name is not None:
            non_tips.append(pos_name)
```

File: empress/tools.py [hunk 13]
```diff
@@ -260,7 +289,7 @@ def match_inputs(
 
         # Subset internal node metadata
         internal_node_names = set([
-            n.name for n in tree.non_tips(include_self=True)
+            name for name in bp_tree_non_tips(bp_tree) if name is not None
```
[the remark this answers is not in the document]
Comment (fedarko) [reply empress/tools.py:265]: Now that these functions don't include `None`s, this line can be replaced with just

```python
internal_node_names = set(bp_tree_non_tips(bp_tree))
```

File: empress/tools.py [hunk 14]
```diff
@@ -144,7 +145,8 @@ def match_inputs(
     # Match table and tree.
     # (Ignore None-named tips in the tree, which will be replaced later on
     # with "default" names like "EmpressNode0".)
-    tip_names = set([n.name for n in tree.tips() if n.name is not None])
+    tip_names = set([name for name in bp_tree_tips(bp_tree)
+                     if name is not None])
```
Comment (fedarko) [inline empress/tools.py:149]:
```suggestion
    tip_names = set(bp_tree_tips(bp_tree))
```

File: empress/tree.py [hunk 15]
```diff
@@ -635,3 +635,63 @@ def update_unrooted_coords(self, s, x1, y1, a, da):
...
+    # Iterate through all open and closing parentheses and extract tip names
+    for i in range(bp_tree.B.size):
+        # Check if this is a leaf node and has a label
+        if isleaf(bp_tree, i) and (bp_tree.name(i) is not None):
+            tips.append(bp_tree.name(i))
+    return tips
+
+
+def bp_tree_non_tips(bp_tree):
+    """ Extracts internal node names in the tree
```
Comment (fedarko) [inline empress/tree.py:662]:
```suggestion
    """ Extracts internal node names in the tree, ignoring unnamed nodes.
```

File: empress/tree.py [hunk 16]
```diff
@@ -635,3 +635,63 @@ def update_unrooted_coords(self, s, x1, y1, a, da):
             max_y, min_y = max(max_y, y2), min(min_y, y2)
 
         return (max_x, min_x, max_y, min_y)
+
+
+def bp_tree_tips(bp_tree):
+    """ Extracts tip names in the tree
```
Comment (fedarko) [inline empress/tree.py:641]:
```suggestion
    """ Extracts tip names in the tree, ignoring unnamed tips.
```

File: tests/python/test_core.py [hunk 17]
```diff
@@ -286,6 +296,24 @@ def test_filter_unobserved_features_from_phylogeny(self):
...
+        ):
+            _ = Empress(self.tree, self.unrelated_table, self.sample_metadata,
+                        filter_unobserved_features_from_phylogeny=False)
+        # Check that --p-filter-missing-features still doesn't work to override
+        # this, since there are NO matching features at all
+        with self.assertRaisesRegex(
+            tools.DataMatchingError,
+            "No features in the feature table are present as tips in the tree."
+        ):
+            _ = Empress(self.tree, self.unrelated_table, self.sample_metadata,
```
Comment (fedarko) [inline tests/python/test_core.py:314]:
```suggestion
            Empress(self.tree, self.unrelated_table, self.sample_metadata,
```

File: tests/python/test_core.py [hunk 18]
```diff
@@ -286,6 +296,24 @@ def test_filter_unobserved_features_from_phylogeny(self):
...
         self.assertIsNone(viz.ordination)
 
+    def test_no_intersection_between_tree_and_table(self):
+        bad_table = self.unrelated_table.copy()
+        bad_table.index = range(len(self.unrelated_table.index))
+        with self.assertRaisesRegex(
+            tools.DataMatchingError,
+            "No features in the feature table are present as tips in the tree."
+        ):
+            _ = Empress(self.tree, self.unrelated_table, self.sample_metadata,
```
Comment (fedarko) [inline tests/python/test_core.py:306]:
```suggestion
            Empress(self.tree, self.unrelated_table, self.sample_metadata,
```

File: tests/python/test_core.py [hunk 19]
```diff
@@ -286,6 +296,24 @@ def test_filter_unobserved_features_from_phylogeny(self):
...
+                        filter_unobserved_features_from_phylogeny=False)
+        # Check that --p-filter-missing-features still doesn't work to override
+        # this, since there are NO matching features at all
```
Comment (fedarko) [inline tests/python/test_core.py:309]:
```suggestion
        # Check that --p-filter-unobserved-features-from-phylogeny doesn't
        # override this: the data mismatch should be identified before attempting
        # shearing
```

File: tests/python/test_core.py [hunk 20]
```diff
@@ -258,6 +296,25 @@ def test_filter_unobserved_features_from_phylogeny(self):
...
 
+    def test_no_intersection_between_tree_and_table(self):
+        bad_table = self.unrelated_table.copy()
+        bad_table.index = range(len(self.unrelated_table.index))
+        with self.assertRaisesRegex(
+            tools.DataMatchingError,
+            "No features in the feature table are present as tips in the tree."
+        ):
+            Empress(self.tree, self.unrelated_table, self.sample_metadata,
+                        filter_unobserved_features_from_phylogeny=False)
```
Comment (fedarko) [inline tests/python/test_core.py:307]:
```suggestion
            Empress(self.tree, self.unrelated_table, self.sample_metadata,
                    filter_unobserved_features_from_phylogeny=False)
```

File: tests/python/test_core.py [hunk 21]
```diff
@@ -258,6 +296,25 @@ def test_filter_unobserved_features_from_phylogeny(self):
...
+        # Check that --p-filter-unobserved-features-from-phylogeny doesn't
+        # override this: the data mismatch should be identified before attempting
+        # shearing
```
Comment (fedarko) [inline tests/python/test_core.py:310]:
```suggestion
        # Check that --p-filter-unobserved-features-from-phylogeny doesn't
        # override this: the data mismatch should be identified before
        # attempting shearing
```

File: tests/python/test_core.py [hunk 22]
```diff
@@ -258,6 +296,25 @@ def test_filter_unobserved_features_from_phylogeny(self):
...
+        with self.assertRaisesRegex(
+            tools.DataMatchingError,
+            "No features in the feature table are present as tips in the tree."
+        ):
+            Empress(self.tree, self.unrelated_table, self.sample_metadata,
+                        filter_unobserved_features_from_phylogeny=True)
```
Comment (fedarko) [inline tests/python/test_core.py:316]:
```suggestion
            Empress(self.tree, self.unrelated_table, self.sample_metadata,
                    filter_unobserved_features_from_phylogeny=True)
```
